src/player.cpp

```diff
@@ -6898,136 +6898,40 @@ bool player::use_charges_if_avail(itype_id it, long quantity)
     return false;
 }
 
-bool player::has_fire(const int quantity) const
+bool player::has_fire( const int quantity ) const
 {
-// TODO: Replace this with a "tool produces fire" flag.
-
-    if( g->m.has_nearby_fire( pos() ) ) {
-        return true;
-    } else if (has_charges("torch_lit", 1)) {
-        return true;
-    } else if (has_charges("battletorch_lit", quantity)) {
-        return true;
-    } else if (has_charges("handflare_lit", 1)) {
-        return true;
-    } else if (has_charges("candle_lit", 1)) {
-        return true;
-    } else if (has_charges("ref_lighter", quantity)) {
-        return true;
-    } else if (has_charges("matches", quantity)) {
-        return true;
-    } else if (has_charges("lighter", quantity)) {
-        return true;
-    } else if (has_charges("flamethrower", quantity)) {
-        return true;
-    } else if (has_charges("flamethrower_simple", quantity)) {
-        return true;
-    } else if (has_charges("hotplate", quantity)) {
-        return true;
-    } else if (has_charges("welder", quantity)) {
-        return true;
-    } else if (has_charges("welder_crude", quantity)) {
-        return true;
-    } else if (has_charges("shishkebab_on", quantity)) {
-        return true;
-    } else if (has_charges("firemachete_on", quantity)) {
-        return true;
-    } else if (has_charges("broadfire_on", quantity)) {
-        return true;
-    } else if (has_charges("firekatana_on", quantity)) {
-        return true;
-    } else if (has_charges("zweifire_on", quantity)) {
-        return true;
-    } else if (has_active_bionic( bio_tools ) && power_level > quantity * 5 ) {
-        return true;
-    } else if (has_bionic( bio_lighter ) && power_level > quantity * 5 ) {
+    if( is_npc() ) {
+        // A hack to make NPCs use their molotovs
         return true;
-    } else if (has_bionic( bio_laser ) && power_level > quantity * 5 ) {
+    } else if( g->m.has_nearby_fire( pos() ) ) {
         return true;
-    } else if( is_npc() ) {
-        // A hack to make NPCs use their Molotovs
+    } else if( has_item_with_flag( "FIRE" ) ) {
         return true;
+    } else if( has_item_with_flag( "FIRESTARTER" ) ) {
```

Comment:
There is no need to check for suitable items first, just iterate over all suitable items directly:
```C++
    ...
    return true;
}
for( item *const i : all_items_with_flag(...) ) {
    if( ... ) {
        return true;
    }
}
return false;
```
If there are no suitable items, the loop will iterate over an empty vector, which means it will be skipped anyway.